File: GMUtilsAndroid/GMUtils/GMUtils/src/main/java/gmutils/firebase/auth/FirebaseAuthManager.java
package gmutils.firebase.auth;

import android.text.TextUtils;

import com.google.firebase.auth.FirebaseAuthException;
import com.google.firebase.auth.FirebaseUser;

import gmutils.StringSet;
import gmutils.firebase.Response;
import gmutils.listeners.ResultCallback;


public class FirebaseAuthManager implements IFirebaseAuthManager {
    public com.google.firebase.auth.FirebaseAuth fbAuth;

    public FirebaseAuthManager() {
        try {
            Class.forName("com.google.firebase.auth.FirebaseAuth");
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            throw new IllegalStateException("add this line to gradle script file:\n" +
                    "//https://firebase.google.com/docs/auth/android/\n" +
                    "implementation 'com.google.firebase:firebase-auth:20.0.2'");
        }

        fbAuth = com.google.firebase.auth.FirebaseAuth.getInstance();
    }

    //----------------------------------------------------------------------------

    @Override
    public void registerByEmail(String email, String password, ResultCallback<Response<Boolean>> callback) {
        fbAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(task -> {
            if (callback != null) {
                if (task.isSuccessful()) {
                    callback.invoke(Response.success(true));
                } else {
                    String ee = "";
                    String ea = "";

                    if (task.getException() != null) {
                        ee = "\nDetails: " + task.getException().getMessage();
                        ea = "\n" + "تفاصيل: " + task.getException().getMessage();
                    }

                    callback.invoke(Response.failed(
                            new StringSet(
                                    "Register failed." + ee,
                                    "تعذر تسجيل الحساب" + ea
                            )
                    ));
                }
            }
        });
    }

    @Override
    public void loginByEmail(String email, String password, ResultCallback<Response<FBUser>> callback) {
        fbAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                FBUser user = convert2FBUser(task.getResult().getUser());
                user.username = task.getResult().getAdditionalUserInfo().getUsername();
                user.extraInfo = task.getResult().getAdditionalUserInfo().getProfile();

                callback.invoke(Response.success(user));
            } else {
                String ee = "";
                String ea = "";
                if (task.getException() != null) {
                    ee = "\nDetails: " + task.getException().getMessage();
                    ea = "\n" + "تفاصيل: " + task.getException().getMessage();
                }
                callback.invoke(Response.failed(
                        new StringSet(
                                "Login failed." + ee,
                                "تعذر تسجيل الدخول" + ea
                        )
                ));
            }
        });
    }

    private FBUser convert2FBUser(FirebaseUser firebaseUser) {
        FBUser user = new FBUser();

        user.email = firebaseUser.getEmail();
        user.displayName = firebaseUser.getDisplayName();
        user.photoUrl = firebaseUser.getPhotoUrl();
        user.phoneNumber = firebaseUser.getPhoneNumber();

        if (firebaseUser.getMetadata() != null) {
            user.creationTimestamp = firebaseUser.getMetadata().getCreationTimestamp();
            user.lastSignInTimestamp = firebaseUser.getMetadata().getLastSignInTimestamp();
        }

        return user;
    }

    @Override
    public FBUser currentUser() {
        FirebaseUser user = fbAuth.getCurrentUser();
        if (user == null) return null;
        return convert2FBUser(user);
    }

    @Override
    public void logout() {
        fbAuth.signOut();
    }

    //----------------------------------------------------------------------------

    @Override
    public void startPasswordReset(String email, ResultCallback<Response<Boolean>> callback) {
        fbAuth.sendPasswordResetEmail(email.trim()).addOnCompleteListener(task -> {
            if (callback != null) {
                if (task.isSuccessful()) {
                    callback.invoke(Response.success(true));
                } else {
                    String ee = "";
                    String ea = "";

                    if (task.getException() != null) {
                        ee = "\nDetails: " + task.getException().getMessage();
                        ea = "\n" + "تفاصيل: " + task.getException().getMessage();
                    }

                    callback.invoke(Response.failed(
                            new StringSet(
                                    "Reset password failed." + ee,
                                    "تعذر تغيير كلمة السر" + ea
                            )
                    ));
                }
            }
        });
    }

    @Override
    public void verifyResetPasswordCode(String email, String resetPasswordCode, ResultCallback<Response<Boolean>> callback) {
        fbAuth.verifyPasswordResetCode(resetPasswordCode).addOnCompleteListener(task -> {
            var emailRelatedCode = task.getResult();
            if (TextUtils.equals(emailRelatedCode, email)) {
                callback.invoke(Response.success(true));
            } else {
                callback.invoke(
                        Response.failed(
                                new StringSet(
                                        "The resetting code is wrong",
                                        "الكود خاطيء"
                                ))
                );
            }
        });

    }

    @Override
    public void changePasswordAfterResetting(String resetPasswordCode, String newPassword, ResultCallback<Response<Boolean>> callback) {
        fbAuth.confirmPasswordReset(resetPasswordCode, newPassword).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                callback.invoke(Response.success(true));
            } else {
                callback.invoke(
                        Response.failed(
                                new StringSet(
                                        "Resetting password failed",
                                        "تعذر تغيير كلمة السر"
                                ))
                );
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public void changePassword(String email, String currentPassword, String newPassword, ResultCallback<Response<Boolean>> callback) {
        if (fbAuth.getCurrentUser() == null) {
            fbAuth.signInWithEmailAndPassword(email, currentPassword).addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    changePassword(fbAuth.getCurrentUser(), newPassword, callback);
                } else {
                    String ee = "";
                    String ea = "";

                    if (task.getException() != null) {
                        ee = "\nDetails: " + task.getException().getMessage();
                        ea = "\n" + "تفاصيل: " + task.getException().getMessage();
                    }

                    callback.invoke(Response.failed(
                            new StringSet(
                                    "Changing password failed." + ee,
                                    "تعذر تغيير كلمة السر" + ea
                            )
                    ));
                }
            });
        } else {
            changePassword(fbAuth.getCurrentUser(), newPassword, callback);
        }
    }

    public void changePassword(FirebaseUser firebaseUser, String newPassword, ResultCallback<Response<Boolean>> callback) {
        firebaseUser.updatePassword(newPassword).addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                callback.invoke(Response.success(true));
            } else {
                String ee = "";
                String ea = "";

                if (task.getException() != null) {
                    ee = "\nDetails: " + task.getException().getMessage();
                    ea = "\n" + "تفاصيل: " + task.getException().getMessage();
                }

                callback.invoke(Response.failed(
                        new StringSet(
                                "Changing password failed." + ee,
                                "تعذر تغيير كلمة السر" + ea
                        )
                ));
            }
        });
    }

    //----------------------------------------------------------------------------

    @Override
    public <T> Response<T> firebaseAuthExceptionMessage(FirebaseAuthException e) {
        String errEn = e.getErrorCode().replace("-", " ");
        String errAr = errEn;

        if (TextUtils.equals(e.getErrorCode(), "network-request-failed")) {
            errEn = "No internet connection";
            errAr = "لا يوجد اتصال بالإنترنت";
            return Response.failed(
                    new StringSet(errEn, errAr),
                    true
            );
        }

        if (TextUtils.equals(e.getErrorCode(), "email-already-in-use")) {
            errEn = IFirebaseAuthManager.EMAIL_ALREADY_IN_USE;
            errAr = "اسم المستخدم مسجل سابقا بالفعل";
            //errAr = "البريد الإلكتروني مستخدم بالفعل";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "invalid-email")) {
            errEn = "Invalid Email";
            errAr = "بريد إلكتروني غير صحيح";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "operation-not-allowed")) {
            errEn = "Operation not allowed";
            errAr = "عملية غير مسموح بها";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "weak-password")) {
            errEn = "Weak password";
            errAr = "كلمة مرور ضعيفة";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "wrong-password")) {
            errEn = "Wrong password";
            errAr = "كلمة مرور خاطئة";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "user-disabled")) {
            errEn = "User disabled";
            errAr = "تم تعطيل الحساب";
        }

        //
        else if (TextUtils.equals(e.getErrorCode(), "user-not-found")) {
            errEn = "User is not found";
            errAr = "الحساب غير موجود";
        }

        if (e.getMessage() != null) {
            errEn += "\n\n<<" + e.getMessage() + ">>";
            errAr += errEn;
        }

        return Response.failed(new StringSet(errEn, errAr));
    }
}

File: GMUtilsFlutter/lib/zgmutils/data_utils/firebase/firebase_auth_manager.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:pallora/zgmutils/utils/string_set.dart';

import '../../utils/logs.dart';
import 'firebase_utils.dart';
import 'response.dart';

abstract class IFirebaseAuthManager {
  static String DEFAULT_EMAIL_HOST_NAME = 'myusers.org';
  static const String anonymousUserName = 'anonymous@anonymous.com';
  static const String anonymousPassword = 'pw0fAnonymous123';

  static Future<UserCredential> anonymousLogin() async {
    var auth = await FirebaseAuthManager().fbAuth;
    var userCredential = await auth.signInWithEmailAndPassword(
      email: anonymousUserName,
      password: anonymousPassword,
    );
    return userCredential;
  }

  //----------------------------------------------------------------------------

  Future<Response<bool>> registerByNonEmail({
    required String text,
    required String password,
  });

  Future<Response<bool>> registerByEmail({
    required String email,
    required String password,
  });

  Future<Response<User>> loginByNonEmail(String text, String password);

  Future<Response<User>> loginByEmail(String email, String password);

  Future<void> logout();

  //----------------------------------------------------------------------------

  String formatNonEmailToEmail(String text) {
    text = FirebaseUtils.refinePhoneNumber(text);
    text = FirebaseUtils.refineKeyName(text);
    return 'un_$text$DEFAULT_EMAIL_HOST_NAME';
  }

  //----------------------------------------------------------------------------

  Future<Response<bool>> startPasswordReset({
    required String identifier,
    required bool isEmailIdentifier,
  });

  Future<Response<bool>> verifyResetPasswordCode({
    required String identifier,
    required bool isEmailIdentifier,
    required String resetPasswordCode,
  });

  Future<Response<bool>> changePasswordAfterResetting(
    String resetPasswordCode,
    String newPassword,
  );

  //----------------------------------------------------------------------------

  Future<Response<bool>> changePassword({
    required String identifier,
    required bool isEmailIdentifier,
    required String currentPassword,
    required String newPassword,
  });

  //----------------------------------------------------------------------------

  static const String EMAIL_ALREADY_IN_USE = 'User identifier already in use';

  Response<T> firebaseAuthExceptionMessage<T>(FirebaseAuthException e);
}

class FirebaseAuthManager extends IFirebaseAuthManager {
  FirebaseAuth? _auth;

  Future<FirebaseAuth> get fbAuth async {
    if (_auth == null) {
      try {
        _auth = FirebaseAuth.instance;
      } catch (e) {
        await Firebase.initializeApp();
        _auth = FirebaseAuth.instance;
      }
    }

    return _auth!;
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<bool>> registerByNonEmail({
    required String text,
    required String password,
  }) {
    return registerByEmail(
      email: formatNonEmailToEmail(text),
      password: password,
    );
  }

  @override
  Future<Response<bool>> registerByEmail({
    required String email,
    required String password,
  }) async {
    try {
      UserCredential userCredential;
      userCredential = await (await fbAuth).createUserWithEmailAndPassword(
        email: email,
        password: password,
      );

      var registered = userCredential.additionalUserInfo?.isNewUser == true;

      return Response.success(data: registered);
    } on FirebaseAuthException catch (e) {
      return firebaseAuthExceptionMessage(e);
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  @override
  Future<Response<User>> loginByNonEmail(String text, String password) {
    return loginByEmail(formatNonEmailToEmail(text), password);
  }

  @override
  Future<Response<User>> loginByEmail(String email, String password) async {
    try {
      UserCredential userCredential;
      userCredential = await (await fbAuth).signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      if (userCredential.user != null) {
        return Response.success(data: userCredential.user);
      } else {
        return Response.failed(
          error: StringSet(
            "login failed",
            "تعذر تسجيل الدخول",
          ),
        );
      }
    } on FirebaseAuthException catch (e) {
      return firebaseAuthExceptionMessage(e);
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  @override
  Future<void> logout() async {
    return (await fbAuth).signOut();
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<bool>> startPasswordReset({
    required String identifier,
    required bool isEmailIdentifier,
  }) async {
    try {
      String email =
          isEmailIdentifier ? identifier : formatNonEmailToEmail(identifier);
      await (await fbAuth).sendPasswordResetEmail(email: email.trim());
      return Response.success(data: true);
    } on FirebaseAuthException catch (e) {
      return firebaseAuthExceptionMessage(e);
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  @override
  Future<Response<bool>> verifyResetPasswordCode({
    required String identifier,
    required bool isEmailIdentifier,
    required String resetPasswordCode,
  }) async {
    try {
      String email =
          isEmailIdentifier ? identifier : formatNonEmailToEmail(identifier);
      var emailRelatedCode = await (await fbAuth).verifyPasswordResetCode(
        resetPasswordCode,
      );
      if (emailRelatedCode != email) {
        return Response.failed(
            error: StringSet(
          'The resetting code is wrong',
          'الكود خاطيء',
        ));
      }

      return Response.success(data: true);
    } on FirebaseAuthException catch (e) {
      return firebaseAuthExceptionMessage(e);
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  @override
  Future<Response<bool>> changePasswordAfterResetting(
    String resetPasswordCode,
    String newPassword,
  ) async {
    try {
      await (await fbAuth).confirmPasswordReset(
        code: resetPasswordCode,
        newPassword: newPassword,
      );
      return Response.success(data: true);
    } on FirebaseAuthException catch (e) {
      return firebaseAuthExceptionMessage(e);
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  //----------------------------------------------------------------------------

  @override
  Future<Response<bool>> changePassword({
    required String identifier,
    required bool isEmailIdentifier,
    required String currentPassword,
    required String newPassword,
  }) async {
    String email =
        isEmailIdentifier ? identifier : formatNonEmailToEmail(identifier);

    var fauth = await fbAuth;

    if (fauth.currentUser == null) {
      await fauth.signInWithEmailAndPassword(
        email: email,
        password: currentPassword,
      );
    }

    try {
      await fauth.currentUser!.updatePassword(newPassword);
      return Response.success(data: true);
    } on FirebaseAuthException catch (e) {
      if (e.code == 'requires-recent-login') {
        try {
          await fauth.signInWithEmailAndPassword(
            email: email,
            password: currentPassword,
          );
          await fauth.currentUser!.updatePassword(newPassword);
          return Response.success(data: true);
        } catch (e) {
          Logs.print(() => e);
          return Response.failed(
              error: StringSet(
            'failed to change password',
            'تعذر تغيير كلمة المرور',
          ));
        }
      } else {
        return firebaseAuthExceptionMessage(e);
      }
    } catch (e) {
      Logs.print(() => e);
      return Response.failed(error: StringSet(e.toString(), e.toString()));
    }
  }

  //----------------------------------------------------------------------------

  @override
  Response<T> firebaseAuthExceptionMessage<T>(FirebaseAuthException e) {
    Logs.print(() => e);

    String errEn = e.code.replaceAll('-', ' ');
    String errAr = errEn;

    if (e.code == 'network-request-failed') {
      errEn = "No internet connection";
      errAr = 'لا يوجد اتصال بالإنترنت';
      return Response.failed(
        error: StringSet(errEn, errAr),
        connectionFailed: true,
      );
    }

    if (e.code == 'email-already-in-use') {
      errEn = IFirebaseAuthManager.EMAIL_ALREADY_IN_USE;
      errAr = 'اسم المستخدم مسجل سابقا بالفعل';
      //errAr = 'البريد الإلكتروني مستخدم بالفعل';
    }

    //
    else if (e.code == 'invalid-email') {
      errEn = 'Invalid Email';
      errAr = 'بريد إلكتروني غير صحيح';
    }

    //
    else if (e.code == 'operation-not-allowed') {
      errEn = 'Operation not allowed';
      errAr = 'عملية غير مسموح بها';
    }

    //
    else if (e.code == 'weak-password') {
      errEn = 'Weak password';
      errAr = 'كلمة مرور ضعيفة';
    }

    //
    else if (e.code == 'wrong-password') {
      errEn = 'Wrong password';
      errAr = 'كلمة مرور خاطئة';
    }

    //
    else if (e.code == 'user-disabled') {
      errEn = 'User disabled';
      errAr = 'تم تعطيل الحساب';
    }

    //
    else if (e.code == 'user-not-found') {
      errEn = 'User is not found';
      errAr = 'الحساب غير موجود';
    }

    if (e.message != null) {
      errEn += '\n\n<<${e.message}>>';
      errAr += '\n\n<<${e.message}>>';
    }

    return Response.failed(error: StringSet(errEn, errAr));
  }
}
